src/QsCompiler/CompilerOptimizations/FunctionEvaluation.fs

```diff
@@ -6,153 +6,147 @@ open Microsoft.Quantum.QsCompiler.DataTypes
 open Microsoft.Quantum.QsCompiler.SyntaxTokens
 open Microsoft.Quantum.QsCompiler.SyntaxTree
 
+open ComputationExpressions
+open TransformationState
 open Utils
 open Printer
 
 
-/// The current state of a function evaluation
-type private FunctionState =
-| Normal // The function is proceeding normally
+/// An interrupt that stops normal function evaluation
+type private FunctionInterrupt =
 | Returned of Expr // The function is returning a value
 | Failed of Expr // The function is throwing an exception
 | CouldNotEvaluate of string // I couldn't evaluate the function
-
-/// The current state of an expression evaluation
-type private ExpressionResult<'A> =
-| ExprValue of 'A
-| ExprError of string
+/// The current state of a function evaluation
+type private FunctionState = Result<TransformationState, FunctionInterrupt>
 
 
 /// Evaluates functions by stepping through their code
-type [<AbstractClass>] internal FunctionEvaluator(cd: CallableDict) =
+type [<AbstractClass>] internal FunctionEvaluator() =
 
     /// Transforms a BoolLiteral into the corresponding bool
     let castToBool x =
         match x with
-        | BoolLiteral b -> ExprValue b
-        | _ -> "Not a BoolLiteral: " + (printExpr x) |> ExprError
-
-    /// Returns whether a FunctionState is an interrupt
-    let isInterrupt = function Normal -> false | _ -> true
+        | BoolLiteral b -> Ok b
+        | _ -> "Not a BoolLiteral: " + (printExpr x) |> CouldNotEvaluate |> Error
 
     /// The callback for the subroutine that evaluates and simplifies an expression
-    abstract member evaluateExpression: VariablesDict -> TypedExpression -> Expr
+    abstract member evaluateExpression: TransformationState -> TypedExpression -> Expr
 
     /// Evaluates a single Q# statement
```

Comment:
That does not give me any information about why e.g. return and fail statements are tagged as Errors?

Reply:
Error is the case of FunctionState that corresponds to a computation not proceeding normally, but instead getting interrupted, eg. by a return or fail statement. I'll add more documentation in a future PR.